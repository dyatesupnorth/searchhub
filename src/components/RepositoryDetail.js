import React, { Component } from 'react'

class RepositoryDetail extends Component {
  render() {
    return (
        <div>
            <h2>Github Repository Details</h2>
        </div>
      
    )
  }
}

export default RepositoryDetail
