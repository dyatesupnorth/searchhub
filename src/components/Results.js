import React from 'react'
const Results = props => (
    <div className="flex items-center justify-center w-full p-4">
        <div className="overflow-hidden bg-white  max-w-xs w-full shadow-lg  leading-normal">
            {
                props.results.items.length === 0 ? ( <p className="text-center m-4"> No Github Repo's Found</p>
                ) : (
                    props.results.items.map((item) => 
                        <span key = {item.id} className={item === props.selectedItem ? 'bg-blue color-white block group cursor-pointer hover:bg-blue p-4 border-b' : 'bg-white block group cursor-pointer hover:bg-blue p-4 border-b'} onClick={() => props.selectItem(item)}>
                            <p className="font-bold text-lg mb-1 text-black group-hover:text-white">{item.full_name}</p>
                            <p className="text-grey-darker mb-2 group-hover:text-white">
                                Open Issues: <strong>{item.open_issues_count}</strong> <br />
                                Forks: <strong>{item.forks_count}</strong>
                            </p>
                        </span>
                    )
            )
        }
        </div>
    </div>    
)


export default Results
