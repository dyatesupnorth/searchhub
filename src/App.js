import React, { Component } from 'react';
import Results from './components/Results'
import RepositoryDetail from './components/RepositoryDetail'
import ReadMe from './components/ReadMe'
import {isEmpty} from './helpers'
import axios from 'axios'
class App extends Component {
   // Set initial state, instead of constructor, reference 'this.props'
   state = {
      result : {
        items:[]
      },
      loading: false,
      error: null,
      searchTerm: '',
      selectedItem: {},
      readme: {}
  }
  handleSubmit = (e) =>{

    e.preventDefault();
    this.setState({
      loading: true
    });

    axios.get(`https://api.github.com/search/repositories?q=${this.state.searchTerm}`)
      .then(res => {
         // Transform the raw data by extracting the nested items
         const items = res.data.items.map(obj => {
           return obj
         });

         // Update state to trigger a re-render.
         // Clear any errors, and turn off the loading indiciator.
         this.setState({
           result:{items},
           loading: false,
           error: null
         });
      })
      .catch(err => {
        console.log('err: ', err);
        this.setState({
          error: err
        });
      });
  }
  getReadme = (repo) => {
    axios.get(`https://api.github.com/repos/${repo.full_name}/readme`)
      .then(res => {
        this.setState({
          readme: res.data
        });
      })
      .catch(err => {
        console.log('err: ', err);
        this.setState({
          error: err
        });
      });
  }
  
  selectItem = (selectedItem) => {
   
    this.setState({selectedItem})

    /* Tightly coupled here, any time a Repo gets called, it'll fire off a call for the readme. 
     This is ok at the minute because the app is really small but might want to sort out the
     way we call the readme in the future.
     */
    this.getReadme(selectedItem)
  }
  onSearchChange = (e) => {
    const searchTerm = e.target.value;
    this.setState(() => ({searchTerm}));
  }
  render() {
    return (
      <div className="container mx-auto">
        <h1 className="text-center uppercase text-grey my-4">Search...hub</h1>
        {!isEmpty(this.state.error) && <p>{this.state.error.response.data.message}</p>}
        <form onSubmit={(e) => this.handleSubmit(e) } className="w-full mx-auto max-w-md">
          <div className="flex justify-center items-center border-b border-b-2 border-teal py-2">
            <input className="appearance-none bg-transparent border-none w-full text-grey-darker mr-3 py-1 px-2 leading-tight"
            type="text" 
            placeholder="Search for a Github repo..." 
            onChange={this.onSearchChange} 
            value={this.state.searchTerm}
            />
            <button className="flex-no-shrink bg-teal hover:bg-teal-dark border-teal hover:border-teal-dark text-sm border-4 text-white py-1 px-2 rounded">Search</button>
            
          </div>
        </form>
        <div className="flex mb-4">
          <div className="w-1/3 h-12">
          {!isEmpty(this.state.result.items) &&  <Results results={this.state.result} selectedItem={this.state.selectedItem} selectItem={this.selectItem}/>}

          </div>
          {!isEmpty(this.state.selectedItem) &&  <div class="w-2/3 rounded bg-grey-lightest my-4">
            <RepositoryDetail item={this.state.selectedItem}  />
            {!isEmpty(this.state.readme) && <ReadMe readme={this.state.readme} />}
          </div>
          }
        </div>
      </div>
    );
  }
}

export default App
