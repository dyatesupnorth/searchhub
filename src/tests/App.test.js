import React from 'react'
import ReactDOM from 'react-dom'
import App from '../App'
import Results from '../components/Results'
import RepositoryDetail from '../components/RepositoryDetail'
import Enzyme, { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16';
import github from './fixtures/github'
import axios from 'axios';

Enzyme.configure({ adapter: new Adapter() });

test('should render App ', () => {
  const wrapper = shallow(<App />)
  expect(wrapper).toMatchSnapshot();

})
describe('fetches a Github repository by search term', () => {


  it('should fetch the data', () => {

    const wrapper = shallow(<App />);

    // Setup the mock response to be from mock data
    const response = github
    // Mock the axios GET call
    axios.get = jest.fn(() => Promise.resolve(response));
 
    //Set new state to be search term...should be handled by input
    wrapper.setState({ searchTerm: 'react' })

     // Submit form using searchTerm
    wrapper.find('form').simulate('submit', {
      preventDefault: ()=>{
      }
      
    });
    // Check GET called correctly
    expect(axios.get).toBeCalledWith('https://api.github.com/search/repositories?q=react');
  
   });

 });