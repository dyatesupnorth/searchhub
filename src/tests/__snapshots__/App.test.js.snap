// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render App  1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="text-center uppercase text-grey my-4"
>
          Search...hub
</h1>,
        false,
        <form
          className="w-full mx-auto max-w-md"
          onSubmit={[Function]}
>
          <div
                    className="flex justify-center items-center border-b border-b-2 border-teal py-2"
          >
                    <input
                              className="appearance-none bg-transparent border-none w-full text-grey-darker mr-3 py-1 px-2 leading-tight"
                              onChange={[Function]}
                              placeholder="Search for a Github repo..."
                              type="text"
                              value=""
                    />
                    <button
                              className="flex-no-shrink bg-teal hover:bg-teal-dark border-teal hover:border-teal-dark text-sm border-4 text-white py-1 px-2 rounded"
                    >
                              Search
                    </button>
          </div>
</form>,
        <div
          className="flex mb-4"
>
          <div
                    className="w-1/3 h-12"
          />
</div>,
      ],
      "className": "container mx-auto",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Search...hub",
          "className": "text-center uppercase text-grey my-4",
        },
        "ref": null,
        "rendered": "Search...hub",
        "type": "h1",
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="flex justify-center items-center border-b border-b-2 border-teal py-2"
>
            <input
                        className="appearance-none bg-transparent border-none w-full text-grey-darker mr-3 py-1 px-2 leading-tight"
                        onChange={[Function]}
                        placeholder="Search for a Github repo..."
                        type="text"
                        value=""
            />
            <button
                        className="flex-no-shrink bg-teal hover:bg-teal-dark border-teal hover:border-teal-dark text-sm border-4 text-white py-1 px-2 rounded"
            >
                        Search
            </button>
</div>,
          "className": "w-full mx-auto max-w-md",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="appearance-none bg-transparent border-none w-full text-grey-darker mr-3 py-1 px-2 leading-tight"
                onChange={[Function]}
                placeholder="Search for a Github repo..."
                type="text"
                value=""
/>,
              <button
                className="flex-no-shrink bg-teal hover:bg-teal-dark border-teal hover:border-teal-dark text-sm border-4 text-white py-1 px-2 rounded"
>
                Search
</button>,
            ],
            "className": "flex justify-center items-center border-b border-b-2 border-teal py-2",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "appearance-none bg-transparent border-none w-full text-grey-darker mr-3 py-1 px-2 leading-tight",
                "onChange": [Function],
                "placeholder": "Search for a Github repo...",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Search",
                "className": "flex-no-shrink bg-teal hover:bg-teal-dark border-teal hover:border-teal-dark text-sm border-4 text-white py-1 px-2 rounded",
              },
              "ref": null,
              "rendered": "Search",
              "type": "button",
            },
          ],
          "type": "div",
        },
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="w-1/3 h-12"
/>,
            false,
          ],
          "className": "flex mb-4",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": false,
              "className": "w-1/3 h-12",
            },
            "ref": null,
            "rendered": false,
            "type": "div",
          },
          false,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="text-center uppercase text-grey my-4"
>
            Search...hub
</h1>,
          false,
          <form
            className="w-full mx-auto max-w-md"
            onSubmit={[Function]}
>
            <div
                        className="flex justify-center items-center border-b border-b-2 border-teal py-2"
            >
                        <input
                                    className="appearance-none bg-transparent border-none w-full text-grey-darker mr-3 py-1 px-2 leading-tight"
                                    onChange={[Function]}
                                    placeholder="Search for a Github repo..."
                                    type="text"
                                    value=""
                        />
                        <button
                                    className="flex-no-shrink bg-teal hover:bg-teal-dark border-teal hover:border-teal-dark text-sm border-4 text-white py-1 px-2 rounded"
                        >
                                    Search
                        </button>
            </div>
</form>,
          <div
            className="flex mb-4"
>
            <div
                        className="w-1/3 h-12"
            />
</div>,
        ],
        "className": "container mx-auto",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Search...hub",
            "className": "text-center uppercase text-grey my-4",
          },
          "ref": null,
          "rendered": "Search...hub",
          "type": "h1",
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="flex justify-center items-center border-b border-b-2 border-teal py-2"
>
              <input
                            className="appearance-none bg-transparent border-none w-full text-grey-darker mr-3 py-1 px-2 leading-tight"
                            onChange={[Function]}
                            placeholder="Search for a Github repo..."
                            type="text"
                            value=""
              />
              <button
                            className="flex-no-shrink bg-teal hover:bg-teal-dark border-teal hover:border-teal-dark text-sm border-4 text-white py-1 px-2 rounded"
              >
                            Search
              </button>
</div>,
            "className": "w-full mx-auto max-w-md",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  className="appearance-none bg-transparent border-none w-full text-grey-darker mr-3 py-1 px-2 leading-tight"
                  onChange={[Function]}
                  placeholder="Search for a Github repo..."
                  type="text"
                  value=""
/>,
                <button
                  className="flex-no-shrink bg-teal hover:bg-teal-dark border-teal hover:border-teal-dark text-sm border-4 text-white py-1 px-2 rounded"
>
                  Search
</button>,
              ],
              "className": "flex justify-center items-center border-b border-b-2 border-teal py-2",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "appearance-none bg-transparent border-none w-full text-grey-darker mr-3 py-1 px-2 leading-tight",
                  "onChange": [Function],
                  "placeholder": "Search for a Github repo...",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Search",
                  "className": "flex-no-shrink bg-teal hover:bg-teal-dark border-teal hover:border-teal-dark text-sm border-4 text-white py-1 px-2 rounded",
                },
                "ref": null,
                "rendered": "Search",
                "type": "button",
              },
            ],
            "type": "div",
          },
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="w-1/3 h-12"
/>,
              false,
            ],
            "className": "flex mb-4",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": false,
                "className": "w-1/3 h-12",
              },
              "ref": null,
              "rendered": false,
              "type": "div",
            },
            false,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
